src/sleep/sleep.c

```diff
@@ -40,6 +43,161 @@
+        size_t fiemap_size = 0, result_fiemap_size = 0;
+
+        if (fstat(fd, &statinfo) != 0) {
+                log_debug_errno(errno, "Cannot determine file size (%s)\n %m",
+                                strerror(errno));
+                return -errno;
+        }
+        fiemap_length = statinfo.st_size;
+
+        fiemap = new(struct fiemap, 1);
+        if (!fiemap)
+                return log_oom();
+
+        result_fiemap = new(struct fiemap, 1);
+        if (!result_fiemap)
+                return log_oom();
+
+        /*  XFS filesystem has incorrect implementation of fiemap ioctl and
+         *  returns extents for only one block-group at a time, so we need
+         *  to handle it manually, starting the next fiemap call from the end
+         *  of the last extent
+         */
+        while (fiemap_start < fiemap_length) {
+                zero(*fiemap);
+
+                fiemap->fm_start = fiemap_start;
+                fiemap->fm_length = fiemap_length;
+                fiemap->fm_flags = FIEMAP_FLAG_SYNC;
+
+                /* Find out how many extents there are */
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0)
+                        return -errno;
+
+                /* Nothing to process */
+                if (fiemap->fm_mapped_extents == 0)
+                        break;
+
+                /* Result fiemap have to hold all the extents for the hole file */
+
+                /* Read in the extents */
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
+
+                /* Resize fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(fiemap, fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
+                        return log_oom();
+
+                memzero(fiemap->fm_extents, extents_size);
+                fiemap->fm_extent_count = fiemap->fm_mapped_extents;
+                fiemap->fm_mapped_extents = 0;
+
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0)
+                        return -errno;
+
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (result_extents +
+                                       fiemap->fm_mapped_extents);
+
+                /* Resize result_fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(result_fiemap, result_fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
+                        return log_oom();
+
+                memcpy(result_fiemap->fm_extents + result_extents,
+                       fiemap->fm_extents,
+                       sizeof(struct fiemap_extent) *
+                       fiemap->fm_mapped_extents);
+
+                result_extents += fiemap->fm_mapped_extents;
+
+                /* Highly unlikely that it is zero */
+                if (fiemap->fm_mapped_extents) {
```

Comment:
we try to avoid on relying on C's downgrade-to-bool feature when checking numeric variables. We do that only for stuff that is boolean or boolean-like (or for pointers). Hence, please write this as `if (fiemap->fm_mapped_extens > 0)` or so